### DISCUSSION TITLE
Future of `meltano elt` or `meltano el` (adapt vs deprecate)

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/meltano#6273] [General]
Opening this discussion to explore use cases that might justify a version of `meltano elt` having ongoing support, versus phasing this out in favor of `meltano run`.

Related:

- https://github.com/meltano/meltano/issues/5940
- https://github.com/meltano/meltano/issues/6164

## Option A - Rename `elt` to `meltano el` to make it clear the behavior and focus

This allows us to move multi-step "el+t" operations to `meltano run` while keeping `meltano el` as a command that can more precisely execute niche use cases - similar to `meltano invoke` having support for additional CLI args vs `meltano run`.

Valid use cases for `meltano el` to continue being supported are primarily for edge cases where a user wants a value for the catalog metadata, the state ID, or the select/exclude lists which is _different_ from those settings specified in the environment. Each of these _can and should_ be set at the environment level in most cases, and any time the user diverges from what is is the environment, there also comes some risk of failures and unexpected results.

Continued support for `meltano elt` (or in future `el`) can address those use cases where some ad-hoc or bespoke operation needs to be performed for a special EL execution - and can solve this for cases where the settings are not valid to be written back to the environment settings. 

For use cases where these settings (catalog metadata and select rules, especially) _are_ appropriate for the environment, then we'd recommend writing those into environment-level settings and then using the more stable and more concise `meltano run` syntax.

## Option B - Deprecate entirely

This wouldn't happen until 3.0, but there are value reasons to keep it alive. Specifically, `meltano run` is designed for safety and conciseness - running a tap or target in `meltano run` will always have behavior that is consistent with the environment-level settings, including auto-mapping to the correct `state_id` for that tap+target combination.

## Note re: migrating to new state ID naming convention

With `meltano run`, state IDs (previously called job IDs) are generated automatically for each environment name and tap/target combination. To move, rename, or otherwise alter those states or their IDs, we have introduced the [`meltano state`](https://docs.meltano.com/reference/command-line-interface#state) command. This can be helpful, for instance, if you want to take an existing `elt` workflow and migrate it to be compatible with the auto-generated naming convention which is used by `meltano run`.

**Comment by tayloramurphy** [comment]
@aaronsteers This recently came up in https://github.com/meltano/meltano/issues/6272#issuecomment-1165872738 and spawned https://github.com/meltano/meltano/issues/6289 as well. 

For sure I don't think we're going to do any deprecations until 3.0. So with that assumption we'll need to continue to support it for now. I believe previously we'd talked about 2 options:

* Update it to `meltano el` to make it clear the behavior and focus
* Deprecate

I believe adding:
- https://github.com/meltano/meltano/issues/6289 

to the documentation would help with this analysis since quickly answering that question takes some leg work right now. At a first glance the main things are:

`meltano elt`:
* --catalog
* --state
* --select / --exclude
* --dump
* Passing the EL loader context and environment to the transfomer

(In reviewing the docs I noticed https://github.com/meltano/meltano/issues/6291 as well which might help a bit with this. B/c I don't know off the top of my head if extractors and loaders can specify commands so that some of this may be solved with that!)

I think if there's a reasonable path to bringing those features to `meltano run` in a way that makes sense, I'm open to dropping `meltano elt`. Otherwise I think we could consider renaming it and focusing its functionality.

**Comment by aaronsteers** [comment]
I have updated the summary for this item:

- https://github.com/meltano/meltano/issues/5940

> 
> ## Updated summary (2022-06-28)
> 
> As of merging #6155, `meltano elt` now fails directly and recommends for Windows users to use `meltano run` instead of `meltano elt`, and `meltano elt` will fail when invoked.
> 
> If we deprecate `meltano elt`, no action will be needed here. If we continue to support `meltano elt`, we'll want to eventually refactor so that it uses the same underlying code paths and `ExtractLoadBlock` that `meltano run` uses.
> 
> Future of `meltano elt` is being discussed here:
> 
> - https://github.com/meltano/meltano/discussions/6273
>

**Reply by s7clarke10** [reply]
Hi @aaronsteers , thank you for opening a discussion on the future of `meltano elt` or `meltano el`.

I have some niche cases for some of my pipelines where I need to explicitly set the name for the `state-id`. I have recently added a number of enhancements to tap-s3-csv https://github.com/s7clarke10/pipelinewise-tap-s3-csv@feature/add_table_suffix.

One of the features in my enhancements allows me to specify a s3 bucket location for loading files sent by a provider and overriding the target filename via environment variables. My Airflow Orchestration runs the same tap `tap-s3-csv` at the same time for every provider (with the benefit of loading the data in parallel). Each parallel load has the s3 bucket location set by an environmental variable. To allow me to run the meltano tasks in parallel I need a different `state-id` name to allow Meltano to run a single tap more than once, and to keep track of incremental processing because file modify dates will be different from one provider to another.

Here is an example dag running tap-s3-csv in parallel for each provider using just a single tap-s3-csv entry in my meltano.yml file. I don't want to add an additional inherited tap-s3-csv for every new provider that I add.

![image](https://user-images.githubusercontent.com/84364906/181395188-8c002ab2-6acd-48da-b076-4cc05446e49e.png)

So in summary, I would either like to keep `meltano elt` or a variant named `meltano el` because it allows me to explicitly set the name for the state-id. It is very easy for my Airflow Dag to set the state-id for each invoked instance of meltano.

I also wonder if there is another option. I would be very happy to adopt `meltano run` if there was an option for me to explicitly set the `state-id`. So if no state-id is set then meltano run operates as it does today. If there was an optional state-id parameter set like `meltano elt`, it would use that name. It would allow those with niche use cases the ability to control / easily set the name for the state-id? Obviously there is a caveat of "beware there be dragons" (pun intended) because if you add other things into your composable pipeline there could be some ill effects.

**Reply by s7clarke10** [reply]
PS. My response is related to this discussion here : https://github.com/meltano/meltano/issues/6164 . 

So perhaps a third option is if there was a known environment variable set, it could be added to the state-id for uniqueness as part of `meltano run`. It might however be more complicated from an engineering perspective if this environment variable is not part of the the tap so wonder if it easier just having the ability to override the state-id name?
